url20b0f880a5_jgraph_legacy_jgraph5_tgz-pJ8-org_jgraph_example_HelloWorldJ8/src/com/jgraph/layout/JGraphWordBrowser.java

```diff
@@ -0,0 +1,480 @@
+					new Object[] { vertex });
+			cells.add(vertex);
+
+			// Removes the cells from the model
+			remove(cells.toArray());
+		}
+		return vertex;
+	}
+
+	/**
+	 * Returns the default port for the specified vertex.
+	 */
+	public Object getDefaultPort(Object vertex) {
+		return getChild(vertex, 0);
+	}
+
+	/**
+	 *
+	 * 
+	 */
+	public void add(String word, String wordType) {
+		if (word.endsWith(")")) {
+			int index = word.indexOf(" (");
+			wordType = word.substring(index + 1);
+			word = word.substring(0, index);
+		}
+		try {
+			List page = getContent(word);
+			Iterator it = page.iterator();
+			String type = "";
+			Object vertex = null;
+			while (it.hasNext()) {
+				String line = (String) it.next();
+				if (line.indexOf("<h3>Noun</h3>") >= 0) {
+					type = "(N)";
+					if (wordType.equals(type)) {
+						vertex = addVertex(word + " " + type);
+					} else {
+						vertex = null;
+					}
+				} else if (line.indexOf("<h3>Adjective</h3>") >= 0) {
+					type = "(A)";
+					if (wordType.equals(type)) {
+						vertex = addVertex(word + " " + type);
+					} else {
+						vertex = null;
+					}
+				} else if (line.indexOf("<h3>Verb</h3>") >= 0) {
+					type = "(V)";
+					if (wordType.equals(type)) {
+						vertex = addVertex(word + " " + type);
+					} else {
+						vertex = null;
+					}
+				}
+				if (vertex != null) {
+					if (line.indexOf("<li>") >= 0) {
+						List words = getWords(line);
+						String edge = "";
+						int index = line.lastIndexOf("(") + 1;
+						if (index >= 0) {
+							edge = line.substring(index, line.indexOf(")",
+									index));
+						}
+						int counter = 1;
+						Iterator it2 = words.iterator();
+						while (it2.hasNext()) {
+							String target = it2.next() + " " + type;
+							if (!target.startsWith("S:")
+									&& !target.equals(vertex)) {
+								addEdge(vertex, target, edge + " (" + counter++
+										+ ")");
+								System.out.println(vertex + " -> " + edge
+										+ " -> " + target);
+							}
+						}
+					}
+				}
+			}
+			sendEdgesToBack();
+		} catch (Exception e) {
+			e.printStackTrace();
+		}
+	}
+
+	public List getWords(String line) {
+		List words = new ArrayList();
+		int index = line.indexOf("<a href=\"");
+		while (index > 0) {
+			index = line.indexOf("\">", index) + 2;
+			int end = line.indexOf("</a>", index);
+			String word = line.substring(index, end);
+			words.add(word);
+			index = line.indexOf("<a href=\"", end + 4);
+		}
+		return words;
+	}
+
+	public List getContent(String word) throws IOException {
+		URL url = new URL("http://wordnet.princeton.edu/perl/webwn");
+		URLConnection urlConn = url.openConnection();
+		urlConn.setDoInput(true);
+		urlConn.setDoOutput(true);
+		urlConn.setUseCaches(false);
+		urlConn.setRequestProperty("Content-Type",
+				"application/x-www-form-urlencoded");
+		DataOutputStream printout = new DataOutputStream(urlConn
+				.getOutputStream());
+		String content = "s=" + URLEncoder.encode(word, "UTF-8")
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;word&#39; (index: 0 | type: String). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other methods validating the parameter: &#39;add&#39;. The same parameter type is validated here for example: njr_failed/url20b0f880a5_jgraph_legacy_jgraph5_tgz-pJ8-org_jgraph_example_HelloWorldJ8/src/com/jgraph/example/SerialGraph.java:55. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).